### PR TITLE
usm: kafka: Parse fetch responses

### PR DESCRIPTION
**Pull request by vitkyrka**
### What does this PR do?

Adds parsing of fetch responses in order to get the counts of records fetched.

### Motivation

https://datadoghq.atlassian.net/browse/USMON-818

### Additional Notes

See this page for more details about the implementation and test results:
https://datadoghq.atlassian.net/wiki/spaces/UT/pages/3575776080/Kafka+Fetch+Response+handling

#### Map sizes

For 30K entries which is the default configuration, the size of the new maps is:

kafka_response: 6.92 MiB
kafka_in_flight: 6.23 MB

#### Load test results

system-probe max k8s CPU Usage
system-probe RSS (avg) (1)
system-probe working set (avg) (1)

### CODE REVIEW

**Comment by DanielLavie** [comment]
Please attach the load test results to the PR

File: pkg/network/usm/kafka_monitor_test.go [hunk 1]
```diff
@@ -9,25 +9,31 @@ package usm
...
 
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
 	"github.com/DataDog/datadog-agent/pkg/ebpf/ebpftest"
 	"github.com/DataDog/datadog-agent/pkg/network/config"
 	"github.com/DataDog/datadog-agent/pkg/network/protocols"
 	"github.com/DataDog/datadog-agent/pkg/network/protocols/http"
 	"github.com/DataDog/datadog-agent/pkg/network/protocols/http/testutil"
 	"github.com/DataDog/datadog-agent/pkg/network/protocols/kafka"
 	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+	"github.com/google/uuid"
```
Comment (DanielLavie) [inline pkg/network/usm/kafka_monitor_test.go:36]: Please fix import order

File: pkg/network/usm/kafka_monitor_test.go [hunk 2]
```diff
@@ -511,7 +629,7 @@ func validateProduceFetchCount(t *testing.T, kafkaStats map[kafka.Key]*kafka.Req
 	require.Equal(t, validation.expectedNumberOfProduceRequests, numberOfProduceRequests,
 		"Expected %d produce requests but got %d", validation.expectedNumberOfProduceRequests, numberOfProduceRequests)
 	require.Equal(t, validation.expectedNumberOfFetchRequests, numberOfFetchRequests,
-		"Expected %d produce requests but got %d", validation.expectedNumberOfFetchRequests, numberOfFetchRequests)
+		"Expected %d fetch requests but got %d", validation.expectedNumberOfFetchRequests, numberOfFetchRequests)
```
Comment (DanielLavie) [inline pkg/network/usm/kafka_monitor_test.go:632]: Nice catch :)

**Review by DanielLavie — DISMISSED** [review]

File: pkg/network/ebpf/c/protocols/kafka/defs.h [hunk 3]
```diff
@@ -12,7 +12,11 @@
 
 #define TOPIC_NAME_MAX_STRING_SIZE 80
 
+#define KAFKA_RESPONSE_PARSER_MAX_ITERATIONS 10
+
+#define KAFKA_MAX_ABORTED_TRANSACTIONS 10000
```
[the remark this answers is not in the document]
Comment (vitkyrka) [reply pkg/network/ebpf/c/protocols/kafka/defs.h:22]: Added.

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h [hunk 4]
```diff
@@ -29,38 +30,606 @@ _Pragma( STRINGIFY(unroll(max_buffer_size)) )
...
+
+static __always_inline bool kafka_process_response(kafka_info_t *kafka, struct __sk_buff* skb, skb_info_t *skb_info) {
+    __u32 offset = skb_info->data_off;
+    conn_tuple_t tup = kafka->tup;
+    __u32 orig_offset = offset;
+    kafka_response_context_t *response = bpf_map_lookup_elem(&kafka_response, &tup);
+    if (response) {
+        u32 expected_seq = 0;
+        if (kafka_is_old_tcp_segment(&tup, skb_info, &expected_seq)) {
+            extra_debug("kafka: skip old TCP segment");
```
[the remark this answers is not in the document]
Comment (vitkyrka) [reply pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h]: Created.

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h [hunk 5]
```diff
@@ -29,38 +30,617 @@ _Pragma( STRINGIFY(unroll(max_buffer_size)) )
...
+    RET_LOOP_END = -2,
+};
+
+// TCP segments splits can happen at any point in the response. If a
+// field happens to straddles the segment boundary, we need to read
+// some bytes from the old packet and the rest from the new packet.
+static __always_inline enum parse_result read_with_remainder(kafka_response_context_t *response, const struct __sk_buff *skb,
+                                                             u32 *offset, s32 *val, bool first)
+{
+    if (*offset >= skb->len) {
```
[the remark this answers is not in the document]
Comment (vitkyrka) [reply pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h]: Done.

File: pkg/network/protocols/kafka/protocol.go [hunk 6]
```diff
@@ -107,19 +136,40 @@ func (p *protocol) PreStart(mgr *manager.Manager) error {
 }
 
 // PostStart empty implementation.
-func (p *protocol) PostStart(*manager.Manager) error {
-	return nil
+func (p *protocol) PostStart(mgr *manager.Manager) error {
+	return p.setupInFlightMapCleaner(mgr)
```
[the remark this answers is not in the document]
Comment (vitkyrka) [reply pkg/network/protocols/kafka/protocol.go:139]: Done.

**Comment by guyarb** [comment]
Please add to the PR description the cost of the 2 new maps (assuming 65K entries)

**Review by guyarb — APPROVED** [review]
LGTM

Please add to the PR description the cost of the 2 new maps (assuming 65K entries)

**Comment by vitkyrka** [comment]
Added load test results and map sizes information to the PR description.

**Comment by guyarb** [comment]
Can merge

**Pull request merged**